Add support for MFA in kube proxies in Connect

ravicious opened the pull request:
Part of #35361.

This PR makes it so that Connect calls `IssueUserCertsWithMFA` to issue certs for kube access. Mind you, this PR merely reuses Brian's work on teleterm-specific MFA prompt. The UI part of it is yet to be finished. This means that the security key will blink whenever a tap is needed, but the UI in Connect won't show anything. TOTP MFA is also not supported yet.

The core of this PR is a migration from `ReissueUserCerts` to `IssueUserCertsWithMFA` in `lib/teleterm/clusters/cluster_kubes.go`. This was a sizeable undertaking, as so far gateways in teleterm assumed that certs are stored on disk and need to be loaded from there, which is exactly the opposite of what local proxies in tsh do – they store the certs solely in memory.

When creating a gateway we can no longer accept just `KeyPath` and `CertPath` – kube gateways now accept `Cert`. `daemon.Service.reissueGatewayCerts`, which orchestrates relogin and cert reissue and is passed to gateways as `OnExpiredCert`, has been changed so that it no longer returns just `error` but now it returns `tls.Certificate` too. This means we can almost directly pass it as `alpnproxy.KubeMiddlewareConfig.CertReissuer`.

Database gateways have been keept untouched. They still use `ReissueUserCerts`. While adding MFA support to them seems fairly straightforward (we just need to generalize `DBCertChecker` so that it can be used by both tsh and Connect), I didn't want to jeopardize quarterly goals.

Best reviewed commit by commit of course. Also a reminder that the layers in lib/teleterm go `daemon` -> `clusters` -> `gateway`, so it's probably the best direction to review the code too.

Changelog: Added per-session MFA support for Kubernetes access in Teleport Connect

ravicious committed:
Improve error handling in gateway/kube_test.go

ravicious committed:
Make gatewaytest.mustGenCertSignedWithCA public

It will be needed to generate and then return cert without saving it to
a file.

ravicious committed:
Move AddMetadataToRetryableError to where it's actually needed

Cluster.reissueDBCerts is called in two scenarios:

* During the request to create a db gateway: this request is made by
  the Electron app, so we need to add metadata to let the app know that
  this error can be resolved by relogin.
* When someone attempts to use the proxy but the user cert has expired:
  this happens outside of the Electron app request-response cycle and
  thus adding metadata is unnecessary.

ravicious committed:
Make gatewaytest.MustGenCACert public, accept CA for MustGenAndSaveCert

This will make writing tests where we need to supply either Cert or KeyPath
and CertPath a little bit easier.

ravicious committed:
Kube gateways: Store cert in memory instead of on disk

ravicious commented on lib/teleterm/clusters/cluster_gateways.go:
```diff
@@ -71,7 +71,10 @@ func (c *Cluster) createDBGateway(ctx context.Context, params CreateGatewayParam
-	if err := c.reissueDBCerts(ctx, routeToDatabase); err != nil {
+	err = AddMetadataToRetryableError(ctx, func() error {
+		return trace.Wrap(c.reissueDBCerts(ctx, routeToDatabase))
+	})
```
This is the same thing that I've done in #35355, but this time for db gateways. When working on #35355, I didn't realize that db gateways were using `AddMetadataToRetryableError` this way too.

ravicious commented on lib/teleterm/clusters/cluster_gateways.go:
```diff
@@ -139,18 +145,36 @@ func (c *Cluster) createKubeGateway(ctx context.Context, params CreateGatewayPar
-		return trace.Wrap(c.reissueDBCerts(ctx, db.RouteToDatabase()))
+
+		// db gateways still store certs on disk, so they need to load it after reissue.
+		err = g.ReloadCert()
```
`daemon.Service.reissueGatewayCerts` used to call `Cluster.ReissueGatewayCerts` followed by `gateway.base.ReloadCert`. Calling it for kube gateways no longer makes sense, so the call to `ReloadCert` has been moved here.

Perhaps I should just rename `ReloadCert` to `ReloadCertFromDisk`.

ravicious commented on lib/teleterm/clusters/cluster_kubes.go:
```diff
@@ -111,22 +114,56 @@ type GetKubesResponse struct {
+		return tls.Certificate{}, trace.Wrap(err)
+	}
+
+	// TODO(ravicious): Add the equivalent of checkIfCertsAreAllowedToAccessCluster.
+
+	// TODO(ravicious): Share this logic with issueKubeCert in tool/tsh/common/kube_proxy.go.
+	if !mfaRequired {
+		if err := c.clusterClient.LocalAgent().AddKubeKey(key); err != nil {
+			return tls.Certificate{}, trace.Wrap(err)
+		}
+	}
+	cert, err := key.KubeTLSCert(kubeCluster)
+	if err != nil {
+		return tls.Certificate{}, trace.Wrap(err)
+	}
+	// Set leaf so we don't have to parse it on each request.
+	leaf, err := utils.TLSCertLeaf(cert)
+	if err != nil {
+		return tls.Certificate{}, trace.Wrap(err)
+	}
+	cert.Leaf = leaf
+
+	return cert, nil
```
I'll tackle sharing code between tsh and teleterm within this PR, but I want to write some integration tests for that new lib/teleterm code first.

greedy52 reviewed (commented):
👍 LGTM overall. 

From refactor perspective, I wonder it maybe be easier to remove KeyPath/CertPath/onNewCertFuncs now? Just need to call `d.setDBCert` after  `d.cfg.OnExpiredCert(ctx, d)` in gateway/db.go, and add certificate to `Cluster.reissueDBCerts`? I don't recall why the certs have to be loaded from files initially.

ravicious committed:
Add ctx to testGatewayCertRenewal

If a gateway requires MFA, daemon.Service.CreateGateway can potentially
block forever. To work around this, we want the tests to pass a context
around so that we can set a timeout on the context in MFA tests.

ravicious committed:
Convert test function args into structs

This will make it easier to pass webauthnLogin for tests that need MFA.

ravicious committed:
Refactor handling of mock tshd events service addr

ravicious committed:
Add tests for MFA

ravicious commented on lib/teleterm/clusters/cluster_kubes.go:
```diff
@@ -111,22 +114,56 @@ type GetKubesResponse struct {
+	// TODO(ravicious): Add the equivalent of checkIfCertsAreAllowedToAccessCluster.
+
+	// TODO(ravicious): Share this logic with issueKubeCert in tool/tsh/common/kube_proxy.go.
+	if !mfaRequired {
+		if err := c.clusterClient.LocalAgent().AddKubeKey(key); err != nil {
+			return tls.Certificate{}, trace.Wrap(err)
+		}
+	}
```
Makes sense, I'm just going to remove it then. I was just copying stuff from tsh without thinking much. It turns out there's not much code to reuse between tsh and teleterm then. Well, excluding the mountains of code we already reuse in the form of local proxies and kube middleware.

Do you know why tsh does this? Is it just to avoid unnecessary cert issues if the cert is already there?

I suppose Connect could do the same as well, but as of today this is not implemented in any gateway.

greedy52 replied:
> Do you know why tsh does this? Is it just to avoid unnecessary cert issues if the cert is already there?

The certs are shared between all `tsh` kube commands. It makes `tsh proxy kube` load quicker but is not necessary IMO. For other commands like `kubectl` (`tsh kube credentials`) or `tsh kubectl`, it is more noticeable whether the certs are reloaded every time or not.

At least my experience with Connect, since there is a GUI where you click things, see tabs opening etc., it doesn't feel slow at all with the current implementation.

ravicious committed:
Do not save kube key in local agent

Connect issues a new cert for every gateway anyway.

ravicious committed:
Check kube certs after issue in teleterm gateways

This makes sure that Connect does the same checks as tsh.

ravicious commented on lib/teleterm/daemon/daemon.go:
```diff
@@ -329,17 +332,25 @@ func (s *Service) reissueGatewayCerts(ctx context.Context, g gateway.Gateway) er
 		}
 
-		if err := cluster.ReissueGatewayCerts(ctx, g); err != nil {
+		promptReason, err := gateway.GetPromptReasonSessionMFA(g.TargetURI())
+		if err != nil {
```
You mean just the struct itself or the whole `GetPromptReasonSessionMFA` function as well? I need to call it both in `daemon` as well as in `clusters` (`lib/teleterm/clusters/gateway_creator.go`), hence why it's defined in `gateway`.

greedy52 replied:
Why does the `GatewayCreator` add `params.MFAPrompt` instead of adding in the daemon before calling the `GatewayCreator`?

ravicious replied:
There was actually a bug in the code that I just solved. `daemon.Service.ResolveClusterURI` attaches an MFA prompt constructor that uses teleterm's MFA prompt. The constructor is attached to a `TeleportClient` instance.

However, `GatewayCreator` doesn't call `daemon.Service.ResolveClusterURI`, it calls `clusters.Storage.ResolveCluster` which is one level below `ResolveClusterURI` and does not set up the MFA constructor.

This meant that when creating a gateway, we weren't actually using the new prompt. 🙈 This also explains why I couldn't verify that `PromptMFA` on tshd events service gets called. I owed it to the `WebauthnLogin` being  weird and skipping calls to the prompt, but that's not the case.

I did a couple of things:

* `daemon.Service` now adds `MFAPrompt` to `clusterCreateGatewayParams` before calling `GatewayCreator`, as you suggested. This fixes the problem with lib/teleterm's prompt not being used during gateway creation.
* I refactored `GetPromptReasonSessionMFA` into `WithPromptReasonSessionMFA` and moved it to the daemon package.
* I added a check to the tests which ensures that the Electron app gets notified about the MFA check.

ravicious committed:
reissueDBCerts → reissueGatewayCerts

ravicious commented on lib/teleterm/gateway/db.go:
> From refactor perspective, I wonder it maybe be easier to remove KeyPath/CertPath/onNewCertFuncs now? Just need to call `d.setDBCert` after `d.cfg.OnExpiredCert(ctx, d)` in gateway/db.go, and add certificate to `Cluster.reissueDBCerts`? I don't recall why the certs have to be loaded from files initially.

The only reason we supply paths is that local proxies work on certs in memory but teleterm uses `ReissueUserCerts` which saves them to disk. So when creating a gateway, we have no choice but to load them from files.

I mean, you're right, the db part of things could be further improved. But the whole thing will have to be refactored anyway when we add MFA to db gateways. I'm on a tight schedule and db MFA is not in scope at the moment, so I'd rather leave it for later.

ravicious commented on lib/client/kube/kube.go:
```diff
@@ -0,0 +1,79 @@
+// defined. If not, it returns an error.
+// This is a safety check in order to print a better message to the user even
+// before hitting Teleport Kubernetes Proxy.
+func CheckIfCertsAreAllowedToAccessCluster(k *client.Key, rootCluster, teleportCluster, kubeCluster string) error {
```
I'm not sure if this is the best place for this function, so please advise if you have a better idea.

It also makes me squirm that when importing this package, I import it as `kubeclient`, even though it does not contain any kind of kube client at the moment. It's more like just a single util function. But that's because `lib/client` is so big, if it was more targeted then I'm sure there would be a relevant `kubeclient` to put this in.

So I'm not sure if I should name the packages according to current meaning or future intent here.

There's `lib/kube/utils` already, so I tried to put this function there, but I ended up with a cyclic import.

Maybe I should just import it as `clientkube` instead?

ravicious commented on lib/client/kube/kube.go:
```diff
@@ -0,0 +1,79 @@
+	}
+	errMsg := "Your user's Teleport role does not allow Kubernetes access." +
+		" Please ask cluster administrator to ensure your role has appropriate kubernetes_groups and kubernetes_users set."
+	return trace.AccessDenied(errMsg)
```
Since this is an access denied error, it won't get shown in the UI in Connect because lib/teleterm truncates all messages from AccessDenied errors (#32550). It'll be visible in logs though and once we address #32550, it'll make its way to the UI as well.

greedy52 reviewed (approved):
Tested with both per-session MFA true and false. Work as expected.

ravicious committed:
Pass mfaPrompt from daemon to GatewayCreator

This also fixes a problem where the mfaPrompt passed from GatewayCreator
was not the expected one. daemon.Service.ResolveClusterURI attaches
MFAPromptConstructor on clusterClient. However, GatewayCreator does not
call daemon.Service.ResolveClusterURI, it calls
clusters.Storage.ResolveCluster which does not attach the proper prompt
constructor.

ravicious committed:
Refactor GetPromptReasonSessionMFA into WithPromptReasonSessionMFA

ravicious committed:
Verify that PromptMFA on tshdEventsService gets called during MFA ceremony

ravicious committed:
reissueKubeCert: Get rid of unused mfaRequired var

ravicious committed:
Use fake storage in gateway CRUD tests

Now that daemon.Service.CreateGateway needs a TeleportClient, we have
to use a fake storage that's able to provide it.

ravicious committed:
Fix synchronization between webauthnLogin and PromptMFA

ravicious commented on integration/proxy/teleterm_test.go:
```diff
@@ -139,17 +139,23 @@ func testGatewayCertRenewal(ctx context.Context, t *testing.T, params gatewayCer
+	webauthLoginCalls := 0
 	webauthnLogin := func(ctx context.Context, origin string, assertion *wantypes.CredentialAssertion, prompt wancli.LoginPrompt, opts *wancli.LoginOpts) (*proto.MFAAuthenticateResponse, string, error) {
 		t.Helper()
+		webauthLoginCalls++
```
I should be fine without using `sync/atomic` here, shouldn't I?

If there was a data race, the detector would catch it. So if someone in the future modifies the MFA code to make concurrent calls for some reason, it'll trigger the data race detector.

ravicious replied:
How wrong I was, I ran the tests myself with `-race` and I already stumbled into this. 🤦

ravicious committed:
Use sync/atomic for function call counts
